src/clarity-angular/input-trap/input-trap.directive.ts

```diff
@@ -0,0 +1,47 @@
+import { Directive, HostListener, ElementRef, OnInit } from "@angular/core";
+
+@Directive({
+    selector: "[clrInputTrap]"
+})
+export class InputTrapDirective implements OnInit {
+
+    static TAB_KEY = 9;
+
+    // tslint:disable-next-line:max-line-length
+    // reference: https://github.com/jkup/focusable
+    static FOCUSABLE_INPUTS = "a[href], input:not([disabled]), select:not([disabled]), textarea:not([disabled]), button:not([disabled])";
+
+    firstInput: HTMLElement;
+    lastInput: HTMLElement;
+
+    constructor(public elementRef: ElementRef) { }
+
+    @HostListener("document:keydown", ["$event"])
+    onkeydown(event: any) {
```

Comment:
We talked about this a couple weeks ago, and the solution we wanted to go with was to listen to the `focusin` event on the whole document, rather than `keydown`. See https://developer.mozilla.org/en-US/docs/Web/Events/focusin.

The keydown solution you have implemented doesn't support `tabindex` (it even breaks if the last element in the DOM of the trap is _not_ the last one in the tab order), it tries to tab to elements even if they are not displayed, it doesn't work when the user interacts manually outside of the trap like the URL bar, ... The focusin solution is far less hackier and much more robust.